llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -495,24 +495,29 @@ class VSETVLIInfo {
   void setAVLIgnored() { State = AVLIsIgnored; }
 
   bool hasAVLImm() const { return State == AVLIsImm; }
-  bool hasAVLReg() const { return State == AVLIsReg; }
+  bool hasAVLDefMI() const { return State == AVLIsDefMI; }
   bool hasAVLVLMAX() const { return State == AVLIsVLMAX; }
   bool hasAVLIgnored() const { return State == AVLIsIgnored; }
   Register getAVLReg() const {
-    assert(hasAVLReg());
-    return AVLReg;
+    assert(hasAVLDefMI());
+    return AVLDefMI->getOperand(0).getReg();
```

Comment:
If the AVL comes from the def of a fault first load then I think it will come from the first operand. I think we might need to track the operand number then in that case?

Reply:
How about keeping both AVLReg and AVLDefMI inside VSETVLInfo? Tracking the operand number looks too indirect because we just want AVLReg here.

`RISCVInsertVSETVLI::insertVSETVLI` is only place need AVLReg actually.

```
  Register AVLReg = Info.getAVLReg();
  MRI->constrainRegClass(AVLReg, &RISCV::GPRNoX0RegClass);
  BuildMI(MBB, InsertPt, DL, TII->get(RISCV::PseudoVSETVLI))
      .addReg(RISCV::X0, RegState::Define | RegState::Dead)
      .addReg(AVLReg)
      .addImm(Info.encodeVTYPE());
```

Reply:
Yeah that also works, as long as we can fit it inside the existing union, something like this maybe?

```c++
  struct Def { const MachineInstr *DefMI; Register Reg; };
  union {
    Def AVLDef;
    unsigned AVLImm;
  };
```

Would need to also update the == and AVL comparison methods to check the register too I think

Reply:
It would be nice to use std::pair here but it has a non-trivial copy constructor unfortunately